### ISSUE TITLE
Some logs overwrites dmeg, grep kern.log (and possible solution)

### ISSUE THREAD

**Issue by ikkaro** [speed47/spectre-meltdown-checker#91]
Hi,
Some logs overwrites the kernel log output. (for intance some iptables logs) and you get a false positive about Meltdown PTI. Can you add this lines of code to grep /var/log/kern.log ? (Lines 1054 to 1057)
```
1050                         elif [ -r /var/log/dmesg ] && grep -Eq "$dmesg_grep" /var/log/dmesg; then
1051                                 # if we can't find the flag in dmesg output, grep in /var/log/dmesg when readable
1052                                 _debug "kpti_enabled: found hint in /var/log/dmesg: "$(grep -E "$dmesg_grep" /var/log/dmesg)
1053                                 kpti_enabled=1
1054                         elif [ -r /var/log/kern.log ] && grep -Eq "$dmesg_grep" /var/log/kern.log; then
1055                                 # if we can't find the flag in dmesg output, grep in /var/log/kern.log when readable
1056                                 _debug "kpti_enabled: found hint in /var/log/kern.log: "$(grep -E "$dmesg_grep" /var/log/kern.log)
1057                                 kpti_enabled=1
1058                         else
1059                                 _debug "kpti_enabled: couldn't find any hint that PTI is enabled"
1060                                 kpti_enabled=0

```

**Comment by speed47** [comment]
I can see that's the case under Debian at least indeed. Patching that.

**Comment by li1ju** [comment]
imho, it can be something else to watch _after_ dmesg, but kern.log is usually rotated after a few days. so you loose the boot information if you look inside kern.log after a certain amount of time :-/

**Comment by speed47** [comment]
Yep, that's why I think I'll have to implement some smarter way to check the dmesg logs :/
Because then just zgrepping /var/log/dmesg.*.gz might find results that don't have anything to do with the current running kernel. I'll come up with a function that does this in a smart-ish way.

**Comment by speed47** [comment]
Grepping in /var/log/dmesg or /var/log/kern.log was causing false positives (we have no guarantee we're grepping messages from the same kernel we're currently running on). To avoid reporting false information to the user, just use `dmesg` and add detection for when it's truncated, and report it.

> commit 40381349ab54208b3bb26dac913b54b8706184f3
> Date:   Sun Jan 21 15:17:10 2018 +0100
> 
>     fix(dmesg): detect when dmesg is truncated
> 
>     To avoid false negatives when looking for a message
>     in dmesg, we were previously also grepping in known
>     on-disk archives of dmesg (dmesg.log, kern.log).
>     This in turn caused false positives because we have no
>     guarantee that we're grepping the dmesg of the current
>     running kernel. Hence we now only look in the live
>     `dmesg`, detect if it has been truncated, and report
>     it to the user.
